Implement the tool (#1)
silverstripe/deprecation-checker: README.md

# Deprecation Changelog Generator

This tool generates the deprecation notice section of a changelog for a Silverstripe CMS release.

It also lets you know if you need to perform follow-up actions (e.g. if some removed API didn't have a deprecation notice).

**This tool is only intended for use by Silverstripe core committers or the Silverstripe Ltd CMS Squad**

## Setup

1. Clone this repo somewhere.
1. Run `composer install` to install its dependencies.

> 1. Set up a directory somewhere that this tool will use for cloning Silverstripe CMS into and for outputting any long-form data.

Reviewer: This really feels like it should be a `_data` style dir that's automatically created and .gitignored, similar to module-standardiser, rather than something you manually created and pass in as an argument

Reviewer 2: I prefer the flexibility to have the output go wherever the heck I want it. If you want it directly in this dir, you can just have this dir as your current working dir and omit the `--dir` optional flag.

I'll update this note to be less "make a new directory" and more "choose where you want to be doing this"

## Usage

If you're unsure of usage at any time, use the `--help` flag to get more details about any command, or run the `bin/deprecation-checker list` command to see all available commands.

> [!NOTE]
> All of the commands accept a `--dir` option. This determines the directory that will be used for all output and must be the same directory across all commands.

> > If this flag is ommitted, the current working directory will be used.

Reviewer: ```suggestion
> If this flag is omitted, the current working directory will be used.
```

Reviewer 2: Done

1. Run `bin/deprecation-checker clone <recipe> <fromConstraint> <toConstraint>`

>     - For example `bin/deprecation-checker clone sink 5.4.x-dev 6.0.x-dev --dir=~/dump/deprecation-checks`

Reviewer: `sink`

Need to document CloneCommand::RECIPE_SHORTCUTS in the readme, or show the command to list the shortcuts in the readme

Reviewer 2: I kinda feel like "If you're unsure of usage at any time, use the `--help` flag to get more details about any command" covers this - but I'll add an explicit note about it.

1. Run `bin/deprecation-checker generate`
    - For example `bin/deprecation-checker generate --dir=~/dump/deprecation-checks`
1. Run `bin/deprecation-checker print-actions`
    - For example `bin/deprecation-checker print-actions --dir=~/dump/deprecation-checks`

This tool uses the `composer` binary on your machine directly, so you shouldn't have any trouble with hitting API rate limits, etc.

## Limitations

> - Due to https://github.com/code-lts/doctum/issues/76 types may not evaluate completely to FQCN
> - The parsing library used doesn't currently pick up on [enums](https://www.php.net/manual/en/language.types.enumerations.php)
> - The parsing library used doesn't currently pick up on globally-scoped consts
> - The parsing library used doesn't currently handle types on constants
> - The parsing library used detects readonly PHPDocs but not the `readonly` keyword for classes and properties

Reviewer: ```suggestion
The parsing library used does not:
- Evaluate types completely to FQCN due to [this issue](https://github.com/code-lts/doctum/issues/76)
- Pick up on [enums](https://www.php.net/manual/en/language.types.enumerations.php)
- Pick up on globally-scoped constants
- Handle types on constants
- Detect the `readonly` keyword for classes and properties (it only detects readonly PHPDocs)
```

Reviewer 2: That doesn't really work with the other items on this list which are not about the parser.... I've reworked this a bit. I also kept the URL in plain text for the issue so it's clear just at a glance that it's an issue in doctum and not in our code.

> - This tool doesn't check YAML files for changes to default values or file/fragment names even through these are in our [definition of public API](https://docs.silverstripe.org/en/project_governance/public_api/)
> - This tool doesn't check calls to `$this->extend()` even through this is in our [definition of public API](https://docs.silverstripe.org/en/project_governance/public_api/)
> - It's not feasible to get config from `get_extra_config()` for comparing.

Reviewer: ```suggestion
This tool also has the following limitations:
- It doesn't check YAML files for changes to default values or file/fragment names, even though these are in our [definition of public API](https://docs.silverstripe.org/en/project_governance/public_api/)
- It doesn't check calls to `$this->extend()`, even though this is in our [definition of public API](https://docs.silverstripe.org/en/project_governance/public_api/)
- It's not feasible to get config from `get_extra_config()` for comparison.
```

Reviewer: Note there was a typo fix in the reformat too - "even through" -> "even though"

Reviewer 2: I've reworked it in a different way because I hadn't seen this when I was looking at the above comment. If you don't like it I can change to what you specifically suggested.

Note that if there's a change hidden in reformatting, it's often easier to see if you make that change in a separate suggestion. Then it will show up in its own diff and often also have its own comment thread for any discussion as needed.
